File: go/dsa/algorithms/sorting/h-index.go
func hIndex(citations []int) int {
	n := len(citations)
	buckets := make([]int, n+1)
	
	for _, c := range citations {
		if c >= n {
			buckets[n]++
		} else {
			buckets[c]++
		}
	}
	
	papersSeen := 0
	for h := n; h >= 0; h-- {
		papersSeen += buckets[h]
		if papersSeen >= h {
			return h
		}
	}
	return 0
}

File: python/dsa/algorithms/sorting/h-index.py
def hIndex(citations):
    n = len(citations)
    buckets = [0] * (n + 1)
    
    for c in citations:
        if c >= n:
            buckets[n] += 1
        else:
            buckets[c] += 1
            
    papers_seen = 0
    for h in range(n, -1, -1):
        papers_seen += buckets[h]
        if papers_seen >= h:
            return h
    return 0